org.osgi.test.common/src/main/java/org/osgi/test/common/context/CloseableBundleContext.java

```diff
@@ -258,19 +262,35 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Throwabl
 					return method.invoke(so, args);
 				}
 			}
+			if (method.getDeclaringClass()
+				.equals(Object.class)) {
+				switch (method.getName()) {
+					case "toString" :
+						return delegatedToString(proxy);
+					case "hashCode" :
+						return so.hashCode();
+					case "equals" :
+						return so.equals(args[0]);
+				}
+			}
 
 			throw new IllegalArgumentException();
 		}
 
+		public String delegatedToString(Object proxy) {
+			return "CloseableServiceObjects[" + System.identityHashCode(proxy) + "]:" + so.toString();
+		}
+
 		@SuppressWarnings("unused")
 		public S getService() {
 			S service = so.getService();
-			instances.add(service);
+			instances.merge(service, 1, (oldValue, dummy) -> oldValue + 1);
 			return service;
 		}
 
 		@SuppressWarnings("unused")
 		public void ungetService(S service) {
+			// FIXME: need to decrement instance count rather than remove completely
```

Comment:
Shouldn't this PR include this?

Reply:
I was planning to do a separate one. There is no code in the BundleContextExtension tests that calls ungetService() nor is there any need for one. I think the proper place for the test is in the commons bundle, which to my mind is a different PR (this one already had too much in it).

But I can look at this.